internal/yggdrasil/yggdrasil_test.go

```diff
@@ -586,6 +592,643 @@ var _ = Describe("Yggdrasil", func() {
 			Expect(eventsRecorder.Events).To(Receive(ContainSubstring("Auth file secret")))
 		})
 
+		It("Secrets reading failed", func() {
+			// given
+			deviceName := "foo"
+			device := getDevice(deviceName)
+			device.Status.Deployments = []v1alpha1.Deployment{{Name: "workload1"}}
+
+			edgeDeviceRepoMock.EXPECT().
+				Read(gomock.Any(), deviceName, testNamespace).
+				Return(device, nil).
+				Times(1)
+
+			secretName := "test"
+			secretNamespacedName := types.NamespacedName{Namespace: device.Namespace, Name: secretName}
+			podData := v1alpha1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "test",
+							Image: "test",
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: secretName,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			deploymentData := &v1alpha1.EdgeDeployment{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      "workload1",
+					Namespace: "default",
+				},
+				Spec: v1alpha1.EdgeDeploymentSpec{
+					DeviceSelector: &v1.LabelSelector{
+						MatchLabels: map[string]string{"test": "test"},
+					},
+					Type: "pod",
+					Pod:  podData,
+					Data: &v1alpha1.DataConfiguration{},
+				}}
+			deployRepoMock.EXPECT().
+				Read(gomock.Any(), "workload1", testNamespace).
+				Return(deploymentData, nil)
+			k8sClient.EXPECT().
+				Get(gomock.Any(), secretNamespacedName, gomock.Any()).
+				Return(fmt.Errorf("test"))
+
+			// when
+			res := handler.GetDataMessageForDevice(context.TODO(), params)
+
+			// then
+			Expect(res).To(Equal(operations.NewGetDataMessageForDeviceInternalServerError()))
+		})
+
+		It("Secrets missing secret", func() {
+			// given
+			deviceName := "foo"
+			device := getDevice(deviceName)
+			device.Status.Deployments = []v1alpha1.Deployment{{Name: "workload1"}}
+
+			edgeDeviceRepoMock.EXPECT().
+				Read(gomock.Any(), deviceName, testNamespace).
+				Return(device, nil).
+				Times(1)
+
+			secretName := "test"
+			secretNamespacedName := types.NamespacedName{Namespace: device.Namespace, Name: secretName}
+			podData := v1alpha1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "test",
+							Image: "test",
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: secretName,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			deploymentData := &v1alpha1.EdgeDeployment{
```

Comment:
The same as podData, can be an external function like getDevice()

Reply:
all tests set deploymentData within the test. this is because, unlike divice object, this is the heart of the tests. Not good to take it out.

Reply:
is all similar, I'm trying to make it simpler to read and not duplicating code. As other parts of the test.